ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java

```diff
@@ -96,21 +97,54 @@ private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> s
     boolean isCopyError = false;
     List<Path> pathList = Lists.transform(srcFileList, ReplChangeManager.FileInfo::getEffectivePath);
     while (!pathList.isEmpty() && (repeat < MAX_COPY_RETRY)) {
-      LOG.info("Attempt: " + (repeat+1) + ". Copying files: " + pathList);
       try {
-        isCopyError = false;
+        // if its retrying, first regenerate the path list.
+        if (repeat > 0) {
+          pathList = getFilesToRetry(sourceFs, srcFileList, destinationFs, destination, isCopyError);
+          if (pathList.isEmpty()) {
+            // all files were copied successfully in last try. So can break from here.
+            break;
+          }
+        }
+
+        LOG.info("Attempt: " + (repeat+1) + ". Copying files: " + pathList);
+
+        // if exception happens during doCopyOnce, then need to call getFilesToRetry with copy error as true in retry.
+        isCopyError = true;
         doCopyOnce(sourceFs, pathList, destinationFs, destination, useRegularCopy);
+
+        // if exception happens after doCopyOnce, then need to call getFilesToRetry with copy error as false in retry.
+        isCopyError = false;
       } catch (IOException e) {
         // If copy fails, fall through the retry logic
-        isCopyError = true;
+        LOG.info("file operation failed", e);
+
+        if (repeat >= (MAX_COPY_RETRY - 1)) {
+          //no need to wait in the last iteration
+          break;
+        }
+
+        if (!(e instanceof FileNotFoundException)) {
+          int sleepTime = FileUtils.getSleepTime(repeat);
+          LOG.info("Sleep for " + sleepTime + " milliseconds before retry " + (repeat+1));
+          try {
+            Thread.sleep(sleepTime);
+          } catch (InterruptedException timerEx) {
+            LOG.info("sleep interrupted", timerEx.getMessage());
+          }
+
+          // looks like some network outrage, reset the file system object and retry.
+          FileSystem.closeAllForUGI(Utils.getUGI());
+          sourceFs = pathList.get(0).getFileSystem(hiveConf);
```

Comment:
need to get new instance instead of getFileSystem call which may get from cache.

Reply:
FileSystem.closeAllForUGI(Utils.getUGI());
sourceFs = pathList.get(0).getFileSystem(hiveConf);
this will work fine